Ignore sources on memberships changing

chrismytton opened the pull request:
If the only thing about a membership that's changing is the source then we want to simply ignore this rather than generating lots of noise reporting on it.

Initial step towards #12 

# Notes to merger

Once https://github.com/everypolitician/everypolitician-pull_request/pull/21 has been merged into `master` we should merge `master` into this pull request to check that it does indeed have regression tests.

chrismytton committed:
Ignore sources on memberships changing

If the only thing about a membership that's changing is the source then
we want to simply ignore this rather than generating lots of noise
reporting on it.

tmtmtmtm commented on lib/everypolitician/pull_request/report/memberships.rb:
```diff
@@ -13,11 +13,16 @@ def removed
+
+        def document_without_sources(doc)
+          doc.delete(:sources)
+          doc
```
This doesn't feel like a very maintainable approach. Extending this to also ignore some other field(s) seems like something that's likely enough that we shouldn't need to rename a method (or chain a lot of similar methods together). We don't really want to over-engineer this either, so I'd be tempted to simply extend the document here with `Hash#except` method from ActiveSupport (or a suitable clone thereof) instead.

chrismytton replied:
Yep, good points! I've added `Hash#except` in https://github.com/everypolitician/everypolitician-pull_request/pull/22/commits/aefe8420a608a5f4b5758f92b5a5fc3198b86f85 and switched to using it in https://github.com/everypolitician/everypolitician-pull_request/pull/22/commits/87164fc283069e7eb7a194363d4d387e3b2e9bcf.

chrismytton committed:
Merge branch 'master' into ignore-membership-source-changing

chrismytton committed:
Make ignored membership fields configurable

Rather than hard-coding the field we want to remove in the method name
it's going to be more future proof to allow the list of fields to
exclude to be configurable. This avoids having to change the method name
in the future when we want to add more fields to be excluded.

tmtmtmtm committed:
Inline `filtered_document`

This version has a little more duplication, but the before/after methods
are already highly similar, and having a helper method on this class that
requires `doc` as an argument is a little bit too much feature-envy at
the wrong level of abstraction for me.

tmtmtmtm commented on lib/everypolitician/pull_request/report/memberships.rb:
```diff
@@ -13,16 +13,15 @@ def removed
-        def document_without_sources(doc)
-          doc.delete(:sources)
-          doc
+        def filtered_document(doc, ignore: %i(sources))
```
This method has a little bit too much feature envy for my liking, and at a slightly odd level of abstraction too (i.e. not even on Membership, but on Document). I fear it's largely premature abstraction, but in a subtly wrong way that's going to bite us later, though I'm afraid that's largely just based on gut reaction and intuition, rather than being able to convincingly argue for it.

The over-engineered solution here would probably be to have our own context-specific wrapper or subclass of Membership, but as we've already simply extended Hash, I think I'd just inline both of these and drop `filtered_document` entirely. We end up with a little bit of duplication, but on methods that are already essentially twins anyway, and when we get to the point where that duplication is too much I suspect we'll want to tease them apart at a slightly different level anyway. 

I'll add that as a new commit to see what you think.

tmtmtmtm reviewed (changes requested):
@chrismytton there's a typo in the commit message for aefe8420a608a5f4b5758f92b5a5fc3198b86f85 ("expect" instead of "except".)

I'm also wondering whether giving credit solely in a commit message is enough when lifting code from elsewhere? I'd be tempted to leave a comment in the code itself too.

(OOI are you conceptually against requiring `active_support/core_ext/hash`, or is it just that this one seems small enough to not require a larger dependency?)

I've also added another commit based on the earlier comment, feel free to push back on that, or even drop it, if you disagree…

chrismytton committed:
Add activesupport dependency

We want to use `Hash#except`, so add activesupport.

The alternative to this would be copying the implementation across, but
using activesupport means less code to maintain in this codebase.
